File: src/lex.py
# Libaries
import re;



# Descriptor class
class Descriptor(object):
	class __Flags(object):
		def __init__(self):
			self.NONE = 0;

	class __States(object):
		def __init__(self):
			pass;

	def __init__(self, flags, flag_ignore):
		self.flags = self.__Flags();
		self.flag_names = [ "NONE" ];
		self.flag_ignore = -1;
		self.type_flags = [];
		self.type_names = [];
		self.states = self.__States();
		self.state_checks = [];
		self.state_flags = [];
		self.state_names = None;
		self.on_new = None;

		for i in range(len(flags)):
			f = flags[i];
			setattr(self.flags, f, 1 << i);
			self.flag_names.append(f);

		self.flag_ignore = getattr(self.flags, flag_ignore);

	def define_types(self, types):
		i = 0;
		for k in types:
			self.type_flags.append(types[k]);
			self.type_names.append(k);
			setattr(self, k, i);
			i += 1;

	def define_state_names(self, state_names):
		for i in range(len(state_names)):
			setattr(self.states, state_names[i], i);

		# Create names
		self.state_names = state_names;

	def define_state(self, state, state_flags):
		self.state_checks.append(state);
		self.state_flags.append(state_flags);

	def type_to_string(self, token_type):
		if (token_type >= 0 and token_type < len(self.type_names)):
			return self.type_names[token_type];
		return "";

	def flags_to_string(self, flags):
		if (flags == 0):
			return self.flag_names[0];

		s = "";
		f = 0x1;

		for i in range(1, len(self.flag_names)):
			if ((flags & f) != 0):
				if (len(s) > 0):
					s += " | ";
				s += self.flag_names[i];

			f <<= 1;

		return s;

	def state_to_string(self, state):
		if (state >= 0 and state < len(self.state_names)):
			return self.state_names[state];
		return "";



# Token class
class Token(object):
	def __init__(self, text, type, flags, state):
		self.text = text;
		self.type = type;
		self.flags = flags;
		self.state = state;

	@classmethod
	def dummy(cls):
		return cls("", -1, 0, -1);



# Bracketed class
class Bracket(object):
	def __init__(self, before, tid, id, opener):
		self.before = before;
		self.token_id = tid;
		self.id = id;
		self.opener = opener;
		self.other = None;



# Tokenizer class
class Lexer(object):
	def __init__(self, descriptor, text):
		self.descriptor = descriptor;
		self.pos = 0;
		self.text = text;
		self.state = 0;
		self.brackets = [];
		self.bracket_stack = [];
		self.previous = Token.dummy();
		self.previous_actual = self.previous;
		self.token_id = 0;

		if (descriptor.on_new is not None):
			descriptor.on_new(self);

	def bracket_track(self, opener):
		# Bracket matching
		if (opener):
			b = Bracket(self.previous, self.token_id, len(self.brackets), True);
			self.brackets.append(b);

			self.bracket_stack.append(b);
		elif (len(self.bracket_stack) > 0):
			b = Bracket(None, self.token_id, len(self.brackets), False);
			self.brackets.append(b);

			b.other = self.bracket_stack.pop();
			b.other.other = b;
		# else: # syntax error

	def create_token(self, token_type, flags, end):
		token = Token(self.text[self.pos : end], token_type, self.descriptor.type_flags[token_type] | self.descriptor.state_flags[self.state] | flags, self.state);

		self.pos = end;

		if ((token.flags & self.descriptor.flag_ignore) != self.descriptor.flag_ignore):
			self.previous = token;
			self.token_id += 1;

		self.previous_actual = token;
		return token;

	def repr_token(self, token):
		return "Token(text={0:s}, type={1:s}, flags={2:s}, state={3:s})".format(repr(token.text), self.descriptor.type_to_string(token.type), self.descriptor.flags_to_string(token.flags), self.descriptor.state_to_string(token.state));

	def match_tree(self, obj, p, p_max):
		value = None;
		o = obj;

		while (True):
			c = self.text[p];
			if (c not in o): break;
			p += 1;

			f = o[c][0];
			if (f is not None):
				value = [ f , p ];

			o = o[c][1];
			if (o is None or p >= p_max): break;

		# Should return [ flags , end ], or None
		return value;

	def get_token(self):
		# Complete?
		if (self.pos >= len(self.text)): return None;

		checks = self.descriptor.state_checks[self.state];
		p = self.pos;
		i_max = len(checks) - 1;

		# Check formations
		for i in range(i_max):
			c = checks[i];
			m = c[0](self, p);
			if (m is not None):
				t = c[1](self, m[0], m[1]);
				if (t is not None):
					return t;

		# Default type
		p += 1;
		flags = 0;

		c = checks[i_max];
		m = c[0](self, p);
		if (m is not None):
			flags = m[0];
			p = m[1];

		return c[1](self, flags, p);



# Useful functions
def tree(obj):
	obj_new = {};

	for k in obj:
		o = obj_new;
		i_max = len(k) - 1;
		for i in range(i_max):
			c = k[i];
			if (c in o):
				if (o[c][1] is None):
					o[c][1] = {};
			else:
				o[c] = [ None , {} ];

			o = o[c][1];

		c = k[i_max];
		if (c in o):
			o[c][0] = obj[k];
		else:
			o[c] = [ obj[k] , None ];

	return obj_new;

def __check_regex_lambda(self, p, re_pattern):
	m = re_pattern.match(self.text, p);
	if (m is None): return None;
	return [ 0 , m.end() ];
def check_regex(pattern, flags=None):
	f = re.DOTALL | re.U;
	if (flags is not None):
		f |= flags;
	re_pattern = re.compile(pattern, f);
	return (lambda self,p: __check_regex_lambda(self, p, re_pattern));

def __check_string_lambda(self, p, text):
	end = p + len(text);
	if (self.text[p : end] == text):
		return [ 0 , end ];
	return None;
def check_string(text):
	return (lambda self,p: __check_string_lambda(self, p, text));

def check_tree(obj):
	return (lambda self,p: self.match_tree(obj, p, len(self.text)));

def __check_null_lambda(self, p):
	return None;
def check_null():
	return __check_null_lambda;

def create_token(token_type):
	return (lambda self,f,p: self.create_token(token_type, f, p));

def __create_token_change_state_lambda(self, flags, p, token_type, state):
	t = self.create_token(token_type, flags, p);
	self.state = state;
	return t;
def create_token_change_state(token_type, state):
	return (lambda self,f,p: __create_token_change_state_lambda(self, f, p, token_type, state));

def __create_token_change_state_before_lambda(self, flags, p, token_type, state):
	self.state = state;
	return self.create_token(token_type, flags, p);
def create_token_change_state_before(token_type, state):
	return (lambda self,f,p: __create_token_change_state_before_lambda(self, f, p, token_type, state));

def to_regex_class(obj):
	s = [];
	o = {};

	for k in obj:
		for c in k:
			o[c] = True;

	for k in o:
		s.append(k);

	return "".join(s);

File: src/lex.js
var lex = (function () {
	"use strict";

	// Basic environment derection; can be hardcoded if the environment is known
	var lex = (function (module) {
		try {
			if (window) return {};
		}
		catch (e) {}

		return module;
	})(this);



	// Descriptor class
	var Descriptor = function (flags, flag_ignore) {
		this.flags = { "NONE": 0 };
		this.flag_names = [ "NONE" ];
		this.flag_ignore = -1;
		this.type_flags = [];
		this.type_names = [];
		this.states = {};
		this.state_names = null;
		this.state_checks = [];
		this.state_flags = [];
		this.on_new = null;

		var i, f;
		for (i = 0; i < flags.length; ++i) {
			f = flags[i];
			this.flags[f] = 1 << i;
			this.flag_names.push(f);
		}

		this.flag_ignore = this.flags[flag_ignore];
	};
	Descriptor.prototype = {
		constructor: Descriptor,
		define_types: function (types) {
			var i = 0,
				k;
			for (k in types) {
				this.type_flags.push(types[k]);
				this.type_names.push(k);
				this[k] = i++;
			}
		},
		define_state_names: function (state_names) {
			this.state_names = state_names;

			for (var i = 0; i < state_names.length; ++i) {
				this.states[state_names[i]] = i;;
			}
		},
		define_state: function (state, state_flags) {
			this.state_checks.push(state);
			this.state_flags.push(state_flags);
		},
		type_to_string: function (token_type) {
			return this.type_names[token_type] || "";
		},
		flags_to_string: function (flags) {
			if (flags === 0) {
				return this.flag_names[0];
			}

			var s = "",
				f = 0x1,
				i = 1;

			for (; i < this.flag_names.length; ++i) {
				if ((flags & f) !== 0) {
					if (s.length > 0) s += " | ";
					s += this.flag_names[i];
				}

				f <<= 1;
			}

			return s;
		},
		state_to_string: function (state) {
			return this.state_names[state] || "";
		},
	};



	// Token class
	var Token = function (text, type, flags, state) {
		this.text = text;
		this.type = type;
		this.flags = flags;
		this.state = state;
	};
	Token.dummy = function () {
		return new Token("", -1, 0, -1);
	};



	// Bracketed class
	var Bracket = function (before, tid, id, opener) {
		this.before = before;
		this.token_id = tid;
		this.id = id;
		this.opener = opener;
		this.other = null;
	};



	// Tokenizer class
	var Lexer = function (descriptor, text) {
		this.descriptor = descriptor;
		this.pos = 0;
		this.text = text;
		this.state = 0;
		this.brackets = [];
		this.bracket_stack = [];
		this.previous = Token.dummy();
		this.previous_actual = this.previous;
		this.token_id = 0;
		this.extra_flags = 0;

		if (descriptor.on_new !== null) {
			descriptor.on_new.call(this);
		}
	};
	Lexer.prototype = {
		constructor: Lexer,
		bracket_track: function (opener) {
			var b;

			// Bracket matching
			if (opener) {
				b = new Bracket(this.previous, this.token_id, this.brackets.length, true);
				this.brackets.push(b);

				this.bracket_stack.push(b);
			}
			else if (this.bracket_stack.length > 0) {
				b = new Bracket(null, this.token_id, this.brackets.length, false);
				this.brackets.push(b);

				b.other = this.bracket_stack.pop();
				b.other.other = b;
			}
			// else: // syntax error
		},
		create_token: function (token_type, flags, end) {
			var token = new Token(this.text.substr(this.pos, end - this.pos), token_type, this.descriptor.type_flags[token_type] | this.descriptor.state_flags[this.state] | this.extra_flags | flags, this.state);

			this.pos = end;

			if ((token.flags & this.descriptor.flag_ignore) !== this.descriptor.flag_ignore) {
				this.previous = token;
				++this.token_id;
			}

			this.previous_actual = token;
			return token;
		},
		repr_token: function (token) {
			return "Token(text=" + lex.repr_string(token.text) + ", type=" + this.descriptor.type_to_string(token.type) + ", flags=" + this.descriptor.flags_to_string(token.flags) + ", state=" + this.descriptor.state_to_string(token.state) + ")";
		},
		match_tree: function (obj, p, p_max) {
			var value = null,
				o = obj,
				c, f;

			while (true) {
				c = this.text[p];
				if (!Object.prototype.hasOwnProperty.call(o, c)) break;
				++p;

				f = o[c][0];
				if (f !== null) {
					value = [ f , p ];
				}

				o = o[c][1];
				if (o === null || p >= p_max) break;
			}

			// Should return [ flags , end ], or null
			return value;
		},
		get_token: function () {
			// Complete?
			if (this.pos >= this.text.length) return null;

			var checks = this.descriptor.state_checks[this.state],
				p = this.pos,
				i = 0,
				i_max = checks.length - 1,
				flags = 0,
				t, c, m;

			// Check formations
			for (; i < i_max; ++i) {
				c = checks[i];
				if ((m = c[0].call(this, p)) !== null && (t = c[1].call(this, m[0], m[1])) !== null) {
					return t;
				}
			}

			// Default type
			++p;
			c = checks[i];
			if ((m = c[0].call(this, p)) !== null) {
				flags = m[0];
				p = m[1];
			}
			return c[1].call(this, flags, p);
		},
	};



	// Useful functions
	var string_repr_escapes = {
		"\b": "\\b",
		"\f": "\\f",
		"\n": "\\n",
		"\r": "\\r",
		"\t": "\\t",
		"\\": "\\\\",
		"\"": "\\\"",
	};
	lex.repr_string = function (s) {
		return '"' + s.replace(/[\b\f\n\r\t\\\"]/g, function (m) { return string_repr_escapes[m]; }) + '"';
	};
	lex.regex_escape = function (text) {
		return text.replace(/[.*+?^${}()|[\]\\\-]/g, "\\$&");
	};
	lex.tree = function (obj) {
		var obj_new = {},
			c, k, o, i, i_max;

		for (k in obj) {
			o = obj_new;
			i_max = k.length - 1;
			for (i = 0; i < i_max; ++i) {
				c = k[i];
				if (Object.prototype.hasOwnProperty.call(o, c)) {
					if (o[c][1] === null) {
						o[c][1] = {};
					}
				}
				else {
					o[c] = [ null , {} ];
				}
				o = o[c][1];
			}

			c = k[i];
			if (Object.prototype.hasOwnProperty.call(o, c)) {
				o[c][0] = obj[k];
			}
			else {
				o[c] = [ obj[k] , null ];
			}
		}

		return obj_new;
	};
	lex.check_regex = function (pattern, flags) {
		flags = (flags === undefined) ? "g" : flags + "g";

		var re = new RegExp("(" + pattern + ")?", flags);
		return function (p) {
			re.lastIndex = p;
			var m = re.exec(this.text);
			return (m[1] === undefined) ? null : [ 0 , m.index + m[0].length ];
		};
	};
	lex.check_string = function (text) {
		return function (p) {
			return (this.text.substr(p, text.length) !== text) ? null : [ 0 , p + text.length ];
		};
	};
	lex.check_tree = function (obj) {
		return function (p) {
			return this.match_tree(obj, p, this.text.length);
		};
	};
	var check_null_fn = function () {
		return null;
	};
	lex.check_null = function () {
		return check_null_fn;
	};
	lex.create_token = function (token_type) {
		return function (flags, p) {
			return this.create_token(token_type, flags, p);
		};
	};
	lex.create_token_change_state = function (token_type, state) {
		return function (flags, p) {
			var t = this.create_token(token_type, flags, p);
			this.state = state;
			return t;
		};
	};
	lex.create_token_change_state_before = function (token_type, state) {
		return function (flags, p) {
			this.state = state;
			return this.create_token(token_type, flags, p);
		};
	};
	lex.to_regex_class = function (obj) {
		var s = "",
			o = {},
			i, k;

		for (k in obj) {
			for (i = 0; i < k.length; ++i) {
				o[k[i]] = true;
			}
		}
		for (k in o) {
			s += k;
		}

		return s;
	};



	// Expose classes
	lex.Descriptor = Descriptor;
	lex.Token = Token;
	lex.Lexer = Lexer;



	// Return module
	return lex;

}).call(this);

